html2docx.py

```diff
@@ -14,138 +15,108 @@ def __init__(self, title: str):
-
-        if self.extra_data:
-            data += self.extra_data
-
-        run = Run(
-            data,
-            bold=bool(self.bold),
-            italic=bool(self.italic),
-            underline=bool(self.underline),
-            strip_whitespace=strip_whitespace,
-        )
-        self.runs.append(run)
-
-    def finish(self) -> None:
-        """Collapse white space across runs."""
-        prev_run = None
-        for run in self.runs:
-            if run.strip_whitespace:
-                run.text = WHITESPACE_RE.sub(" ", run.text.strip())
-                if prev_run and prev_run.strip_whitespace:
-                    if prev_run.needs_space_suffix:
-                        prev_run.text += " "
-                    elif run.text and run.needs_space_prefix:
-                        run.text = " " + run.text
-            prev_run = run
-
-        for run in self.runs:
-            if run.text:
-                doc_run = self.p.add_run(run.text)
-                doc_run.bold = run.bold
-                doc_run.italic = run.italic
-                doc_run.underline = run.underline
-
-        self.p = None
-        self.runs = []
-
-    def add_list_style(self, style: str) -> None:
-        self.finish()
-        if self.list_style:
-            # The template included by python-docx only has 3 list styles.
-            level = min(len(self.list_style) + 1, 3)
-            style = f"{style} {level}"
-        self.list_style.append(style)
-
-
-class Run:
-    def __init__(
-        self,
-        text: str,
-        bold: bool = False,
-        italic: bool = False,
-        underline: bool = False,
-        strip_whitespace: bool = True,
-    ):
-        self.text = text
-        self.bold = bold
-        self.italic = italic
-        self.underline = underline
-        self.strip_whitespace = strip_whitespace
-
-        self.needs_space_prefix = bool(WHITESPACE_RE.match(text[0]))
-        self.needs_space_suffix = bool(WHITESPACE_RE.match(text[-1]))
+        if not data.strip() and self.p is None:  # whitespace between block tags
+            return
+        if not self.pre:
+            data = re.sub(WHITESPACE_RE, " ", data)
+        if self.collapse_space:
+            # https://www.w3.org/TR/css-text-3/#collapse
+            # keep track of `segment breaks` to impose white space collapsing
+            data = data.lstrip()
+        if data:
+            self.collapse_space = data.endswith(" ")
+            self._add_text(data)
+
+    def handle_endtag(self, tag: str) -> None:
+        if tag in ["a", "b", "em", "i", "strong", "u"]:
+            if tag == "a" and self.href:
+                self._add_text(f" {self.href}")
+                self.href = ""
+            self._finish_run()
+        elif tag in ["h1", "h2", "h3", "h4", "h5", "h6", "li", "ol", "p", "pre", "ul"]:
+            self._finish_p()
+            if tag in ["ol", "ul"]:
+                self.list_styles = self.list_styles[:-1]
```

Comment:
```suggestion
                del self.list_styles[-1]
```

Reply:
The precondition to running html2docx is for the HTML to be valid. `<li>item</li></ul>` is not a valid construct.
https://github.com/erezlife/html2docx/blob/58ea8f612958589214d973ad68f60a7082ca3932/html2docx.py#L146

Reply:
Actually I removed this doc string. But the point here is not this.
There is 2 outcomes that this policy may have:
1. Invalid HTML is rejected with a specific exception - not `IndexError`. This means I will have to introduce something like this:
```
if self.list_styles:
    del self.list_styles[-1]
else:
    raise ...
```
And once I introduce this check, then to be consistent I need to validate all HTML (and this code never had this type of validation).
2. Invalid HTML is not rejected but the correct result is not guaranteed (current case).
Please not that we do not want to send 500 response. So we need to handle any case of exception.
Your call.

Reply:
This PR is about refactoring the code, I strongly disagree to the removal of the valid HTML precondition, especially in this PR.
In general, `html2docx` is not meant to validate HTML, that’s the responsibility of the library users. Handling invalid HTML is beyond the scope of this project and I am currently opposed to removing this precondition.

As an example, the erezlife project uses [`bleach`](https://bleach.readthedocs.io/en/latest/) to validate the HTML fragments submitted by users and verify they are well-formed. The bleach library [handles unclosed or mis-nested tags](https://bleach.readthedocs.io/en/latest/clean.html), so erezlife should always pass valid HTML fragments to `html2docx`.

With the valid HTML pre-condition, I believe calling `del self.list_styles[-1]` is safe and won’t trigger an exception. If you have valid HTML that trigger the `IndexError`, please share it.

Reply:
If you read carefully you can notice I am against introducing HTML validation here.
I'm OK if bleach is used to validate the input.

I was only referring to invalid HTML - but you say we should not care about it.